File: python/number-of-valid-words-for-each-puzzle.py
import collections

class Solution:
    def findNumOfValidWords(self, words: List[str], puzzles: List[str]) -> List[int]:
        def get_bit_mask(word):
            mask = 0
            for c in word:
                mask |= 1 << (ord(c) - ord('a'))
            return mask
        
        counter_words = collections.Counter()
        for word in words:
            mask = get_bit_mask(word)
            counter_words[mask] += 1
        
        res = []
        for puzzle in puzzles:
            count = 0
            fst_letter = 1 << (ord(puzzle[0]) - ord('a'))
            mask = get_bit_mask(puzzle)
            sub_mask = mask
            while True:
                if sub_mask & fst_letter:
                    count += counter_words[sub_mask]
                sub_mask = (sub_mask - 1) & mask
                if sub_mask == 0:
                    break
            res.append(count)
        return res

File: go/number_of_valid_words_for_each_puzzle.go
package leetcode

func findNumOfValidWords(words []string, puzzles []string) []int {
	dic_words := make(map[int]int)
	for _, word := range words {
		mask := get_bit_mask(word)
		dic_words[mask] += 1
	}

	res := make([]int, len(puzzles))
	for i, puzzle := range puzzles {
		mask := get_bit_mask(puzzle)
		fst_letter := 1 << int(puzzle[0]-'a')
		count := 0
		subMask := mask
		for {
			if subMask&fst_letter != 0 {
				count += dic_words[subMask]
			}
			subMask = (subMask - 1) & mask
			if subMask == 0 {
				break
			}
		}
		res[i] = count
	}
	return res
}

func get_bit_mask(word string) int {
	res := 0
	for i := 0; i < len(word); i++ {
		res |= 1 << int(word[i]-'a')
	}
	return res
}
